File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:

    bool palindrome(int left, int right, string s){
        while(left < right){
            if(s[left] != s[right]) return false;
            left++; right--;

        }
        return true;
    }

    void func(int ind, string s, vector<string> &ds, vector<vector<string>> &ans, int n){

        // base case:
        if(ind == n){
            ans.push_back(ds);
            return;
        }

        for(int i = ind; i<n; i++){

            if( palindrome(ind,i,s) ){

                string part = "";
                for(int j=ind;j<=i;j++) part += s[j];

                ds.push_back(part);
                func(i + 1, s, ds, ans, n);
                ds.pop_back();
            }
        }

    }

    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> ds;
        int n = s.size();

        func(0,s,ds,ans,n);
        
        return ans;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {


    public List<List<String>> partition(String s) {
        
        List<List<String>> ans = new ArrayList<>();
        List<String> ds = new ArrayList<>();

        func(0,s,ans,ds);
        return ans;
    }

    public void func(int ind, String s, List<List<String>> ans, List<String> ds){

        if(ind == s.length()){
            ans.add(new ArrayList<>(ds));
            return;
        }

        for(int i=ind;i<s.length();i++){

            if(isPalindrome(ind,i,s)){

                ds.add(s.substring(ind,i+1));
                func(i+1,s,ans,ds);
                ds.remove(ds.size() - 1);
            }
        }
    }

    public boolean isPalindrome(int start, int end, String s){
        while(start<end){
            if(s.charAt(start++) !=  s.charAt(end--)) return false;
        }
        return true;
    }
}